webhook/main.go

```diff
@@ -0,0 +1,154 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"go.uber.org/zap/zapcore"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	esv1alpha1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/controllers/crds"
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+const (
+	errCreateWebhook = "unable to create webhook"
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = esv1beta1.AddToScheme(scheme)
+	_ = esv1alpha1.AddToScheme(scheme)
+}
+
+func main() {
+	var metricsAddr string
+	var enableLeaderElection bool
+	var loglevel string
+	var namespace string
+	var dnsName string
+	var certDir string
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&dnsName, "dns-name", "localhost", "DNS name to validate certificates with")
+	flag.StringVar(&certDir, "cert-dir", "/tmp/k8s-webhook-server/serving-certs", "path to check for certs")
+	flag.StringVar(&loglevel, "loglevel", "info", "loglevel to use, one of: debug, info, warn, error, dpanic, panic, fatal")
+	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	flag.StringVar(&namespace, "namespace", "", "watch external secrets scoped in the provided namespace only. ClusterSecretStore can be used but only work if it doesn't reference resources from other namespaces")
+	flag.Parse()
+
+	var lvl zapcore.Level
+	err := lvl.UnmarshalText([]byte(loglevel))
+	if err != nil {
+		setupLog.Error(err, "error unmarshalling loglevel")
+		os.Exit(1)
+	}
+	c := crds.CertInfo{
+		CertDir:  certDir,
+		CertName: "tls.crt",
+		KeyName:  "tls.key",
+		CAName:   "ca.crt",
+	}
+
+	logger := zap.New(zap.Level(lvl))
+	ctrl.SetLogger(logger)
+
+	setupLog.Info("validating certs")
+	err = crds.CheckCerts(c, dnsName, time.Now().Add(time.Hour))
+	if err != nil {
+		setupLog.Error(err, "error checking certs")
+		os.Exit(1)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go func(c crds.CertInfo, dnsName string) {
+		// notify on sigterm and hour
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		ticker := time.NewTicker(time.Hour)
+		for {
+			select {
+			case <-sigs:
+				cancel()
+			case <-ticker.C:
+				setupLog.Info("validating certs")
+				err = crds.CheckCerts(c, dnsName, time.Now().Add(crds.LookaheadInterval+time.Minute))
```

Comment:
I just realized that we need to be sure that we check for certificate validation using the same interval as the crd controller.. Which means we can be `crd-controller ReconcileAfter + 1.minute` of max unavailability... Otherwise it could go up to the ticker time.

I'm not sure this is a good solution, but it would take care of the main issues.

Reply:
It does not, for instance, prevents from someone deleting the certificate content in the secret. The CRD controller will regenerate in `certController.requeueInterval` but the webhook will `still` have a valid certificate around.

To improve that, I think the best approach (for another issue, another PR) is add a watch on that particular secret, and when the watch triggers, we restart the webhook.